cargo/private/cargo_build_script.bzl

```diff
@@ -277,7 +277,16 @@ cargo_build_script = rule(
             allow_files = True,
         ),
         "deps": attr.label_list(
-            doc = "The Rust dependencies of the crate",
+            doc = "The Rust build-dependencies of the crate",
+            providers = [rust_common.dep_info],
+            cfg = "exec",
+        ),
+        "link_deps": attr.label_list(
```

Comment:
Rather than introducing a new attribute on `cargo_build_script`, which requires manual maintenance of this information in BUILD files, can we add a new `links` field to `DepInfo` and `BuildInfo` here: https://github.com/bazelbuild/rules_rust/blob/f1b19c394f3edffa041b88be1bc371950476d1d2/rust/private/providers.bzl#L46-L70? That way we can conditionally inspect `dep_env` based on whether the crate had a `links` attribute, but the BUILD file doesn't need to know about the difference.

Reply:
I am confused as I don't see where I would gather this information from. `link_deps`, and `deps` are disjoint ideas (neither is a subset of the other). One comes from *build* dependencies for the host, and the other comes from *runtime* (normal) dependencies for the target which are not the same. Please let me know if I am misunderstanding you.

Reply:
I primarily made this change as we have link_deps that link against .so libraries that we didn't want to have added to the cargo_build_script as they aren't there in cargo and aren't needed.

Reply:
Do you have a cargo equivalent of what you're doing here? Or an example repo you can share that shows what you're trying to do?

I've been assuming you're basically implementing support for the [`links`](https://doc.rust-lang.org/cargo/reference/build-scripts.html#the-links-manifest-key) attribute in a Cargo.toml file.

The way I think cargo models this is: a build script may optionally propagate some settings to the linker to add configuration to the linker action, But a cargo build script can _also_ do other things like generate code which is needed for compiles, or generate environment variables which should be set for targets which depend on this one.

That's why I'm viewing the "link_deps" behaviour as a subset of the general "depending on a cargo_build_script" behaviour - if a cargo_build_script does both codegen (for code that needs compiling as part of depending on this cargo_build_script) and specifying link libraries, then if we're looking to closely model cargo build scripts (which is the aim of the `cargo_build_script` rule), we should expose this as optional subsets.

If you're just looking to link against an existing .so file, you should be able to just add a dependency from your `rust_library` or `rust_binary` on a [`cc_import`](https://bazel.build/reference/be/c-cpp#cc_import) target (or a [`cc_library`](https://bazel.build/reference/be/c-cpp#cc_library) if you're compiling it yourself) and shouldn't need to go via a `cargo_build_script` at all.

Reply:
> Do you have a cargo equivalent of what you're doing here?

Yes, but a full example is overly complicated. I have attached a simpler example below.

Here is an example of something that compiles in Bazel that doesn't in Cargo (without this PR). This is because of the lack of difference between `deps`, and `link_deps` that isn't made before this PR. 

> I've been assuming you're basically implementing support for the [links](https://doc.rust-lang.org/cargo/reference/build-scripts.html#the-links-manifest-key) attribute in a Cargo.toml file.

I am attempting to fix some details of the `link` support which already exists but aren't implemented properly.

> The way I think cargo models this is: a build script may optionally propagate some settings to the linker to add configuration to the linker action, But a cargo build script can also do other things like generate code which is needed for compiles, or generate environment variables which should be set for targets which depend on this one.

I am aware of that already. I attempted to highlight the more subtle nuances above, but I will repeat them again. `build-dependencies` in a Cargo.toml file are associated with `cargo_build_script` `deps` while `dependencies` in a Cargo.toml file (that have a `links` attribute) are associated with `cargo_build_script` `link_deps`.

> That's why I'm viewing the "link_deps" behaviour as a subset of the general "depending on a cargo_build_script" behaviour - if a cargo_build_script does both codegen (for code that needs compiling as part of depending on this cargo_build_script) and specifying link libraries, then if we're looking to closely model cargo build scripts (which is the aim of the cargo_build_script rule), we should expose this as optional subsets.

Neither is a subset of the other although I could see how the Rust documentation could mislead you into thinking that. I initially thought that was the case at well until reading more.

> If you're just looking to link against an existing .so file, you should be able to just add a dependency from your rust_library or rust_binary on a [cc_import](https://bazel.build/reference/be/c-cpp#cc_import) target (or a [cc_library](https://bazel.build/reference/be/c-cpp#cc_library) if you're compiling it yourself) and shouldn't need to go via a cargo_build_script at all.

I am aware of that. My goal was to *not* link against something which it currently is. This is because `deps` link things while `link_deps` won't. (This was because we are cross compiling, and only have the target architecture version of a library available to the rule.)